hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/BucketIndexBulkInsertPartitioner.java

```diff
@@ -59,12 +62,19 @@ public BucketIndexBulkInsertPartitioner(HoodieTable table, String sortString, bo
       this.sortColumnNames = null;
     }
     this.preserveHoodieMetadata = preserveHoodieMetadata;
+    // Bulk insert into COW table with bucket index is allowed only once, otherwise AppendHadleFactory will produce MOR log files
+    this.isAppendAllowed = !table.getConfig().getTableType().equals(HoodieTableType.COPY_ON_WRITE);
   }
 
   @Override
   public Option<WriteHandleFactory> getWriteHandleFactory(int idx) {
-    return doAppend.get(idx) ? Option.of(new AppendHandleFactory()) :
-        Option.of(new SingleFileHandleCreateFactory(FSUtils.createNewFileId(getFileIdPfx(idx), 0), this.preserveHoodieMetadata));
+    if (!doAppend.get(idx)) {
+      return Option.of(new SingleFileHandleCreateFactory(FSUtils.createNewFileId(getFileIdPfx(idx), 0), this.preserveHoodieMetadata));
+    } else if (isAppendAllowed) {
+      return Option.of(new AppendHandleFactory());
+    } else {
+      throw new HoodieNotSupportedException("Bulk insert into COW table with bucket index is allowed only once, please, use upsert operation instead");
```

Comment:
may overwrite a table use bulk_insert more time? and I think if user need, they can insert into table use bulk_insert more time, because bulk_insert is more effective than other op.

Reply:
Check that overwrite is working properly, and fixed failed tests.

Reply:
@danny0405 , need some time to figure out, how `RDDSimpleBucketBulkInsertPartitioner` should work. Looks like it's not working properly, and possible there is a new separate bug. But it affects how to made fix in this MR.

Reply:
Currently, we create some strange data structures, I don't understand the. For instance, into `fileIdPfxList` we place file names, which exist:
https://github.com/apache/hudi/blob/b31c858556ce783694bfec75dc38d51e58ced993/hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/execution/bulkinsert/RDDSimpleBucketBulkInsertPartitioner.java#L88-L92

And add not existed in the end:
https://github.com/apache/hudi/blob/b31c858556ce783694bfec75dc38d51e58ced993/hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/execution/bulkinsert/RDDSimpleBucketBulkInsertPartitioner.java#L97-L102

And later in `BucketIndexBulkInsertPartitioner`, I should decide what to do with current data from this data:
![cow-as-mor-fileIdPfxList](https://github.com/user-attachments/assets/e2bced79-1990-4fa1-ae56-c228c063281a)

Here I've inserted `record1` with `bucketID = 161`. And at the debug point, I want to insert `record2` with `bucketID = 138`. But incoming `int partitionId = 1`, which came from Spark.

I will try to figure out, how it should work properly.

Reply:
There should be two kinds of "partition id" here, one for Spark partition, another for Hudi bucket id(file group id), the reason to resolve a prefix from the file group id is that: for bucket index, the prefix number string is kind of the bucket id from Hudi insights, the whole file group id is comprised by a number string and an uuid, where here the uuid string is kind of inoperative.

Reply:
I kind of give the wrong info before, I reliazed that we are using bucket index, bucket index is used only for `UPSERT` semantics, because the same keys should always be written into one file group, if we allow multiple bulk_inserts here as inserts, there would be a file slice with multiple base files, this may work but usually we do not do that, the suggested coposition is: bulk_insert (one shot for better performance) + multiple upserts.

Reply:
I suppose we should throw exception then in the case when the second bulk insert try to write to existed bucket.

Reply:
For COW table, I would say yes.